docs/fpc-key-dist.puml

```diff
@@ -1,97 +1,319 @@
-    - S_State denotes state secret key.
-    - CC_{PK,SK} denotes chain code pub/priv key pair
+  note over Admin1, Admin2
+    At this point, a chaincode enclave is up and running, and another is waiting for the chaincode keys.
+    We need a means to trigger the key distribution. Now, an admin does it.
+    Can Fabric automate this?
   end note
 
-  Enclave1 -> Enclave1 : m <- encrypt {{S_State, CC_SK}} with ECC_E_PK
-  note right
-    enclave encryptes state secret key and chaincode privkey.
-    - m denotes encrypted message
+  Admin1        -> ERCC1            ++: query registered enclave's key
+    note right
+        ERCC will likely return the list of enclaves
+        with their cryptographic material. Admin1
+        then selects the non-provisioned one(s).
+    end note
+  return ECC2_S_PK /'Admin1 -> ERCC1'/
+
+  Admin1        -> Admin1           :   m <- <CC_name, CC_ver, CH_id, org_id, ECC2_S_PK>
+  Admin1        -> ECC_Manager1     ++: DistributeKey(m)
+  ECC_Manager1  -> ECC1             ++: DistributeKey(m)
+  ECC1          -> Enclave1         ++: DistributeKey(m)
+  Enclave1      -> ERCC1            ++: GetRegistration(ECC2_S_PK)
+  return registration_message <- <m, attestation_report> /'Enclave1 -> ERCC1'/
+  Enclave1      -> TLCC_Enclave1    ++: check_committed_data(registration_message)
+  note left of TLCC_Enclave1
+    The actual implementation of check_committed_data
+    is out of scope here. A possible solution can be
+    sketched as follows. A principal asks TLCC whether
+    a chaincode (say ERCC) has stored a key
+    (say ECC2_SPK||".registration"), with a specific hash
+    value (say h=H(m || attestation_report)). TLCC returns
+    true/false (say h==H(value of store key)).
   end note
-
-  Enclave1 -> Enclave1 : SIG_ECC <- sign {{m, CC_PK}} with ECC_S_SK
-  note right
-    enclave signs the encrypted message and the chaincode pubkey.
-    - SIG_ECC denotes the signature
-
+  return true /'Enclave1 -> TLCC_Enclave1'/
+    note left
+        if false, abort
+    end note
+  Enclave1      -> TLCC_Enclave1    ++: can_endorse(ECC2_S_PK)
+    note right
+        TLCC checks the endorsement policy to determine
+        if the recipient enclave is an eligible endorser.
+    end note
+    note right
+        TLCC should check that the recipient enclave's MRENCLAVE
+        matches the chaincode definition's MRENCLAVE.
+        This could enable key distribution from an old cc version,
+        with a different MRENCLAVE.
+        Notice: in this case there cannot be also a check
+        can_endorse(ECC1_S_PK) -- or it should be negated.
+    end note
+  return true /'Enclave1 -> TLCC_Enclave1'/
+    note left
+        if false, abort
+    end note
+  Enclave1      -> Enclave1         :   check CC_name, CC_ver, CH_id, org_id in sealed data
+  Enclave1      -> Enclave1         :   check ECC2_S_PK, ECC_E_PK, attestation_report
+  Enclave1      -> Enclave1         :   m <- encrypt <SEK, CC_E_SK> with ECC2_E_PK
```

Comment:
I know I introduced `m` here but I don't like the single letter here. Maybe we call id `SharedCCKeys` or maybe with a `_target` subscript to denote that this is for ECC2?

Reply:
changed m, but I'd rather not tag messages with recipients identities - ECC2 does not know the a message is for it until after the signature verification.